src/controllers/Users/oauth2/discord.ts

```diff
@@ -0,0 +1,215 @@
+import axios from 'axios'
+import { Request, Response, Router } from 'express'
+import { query } from 'express-validator'
+import querystring from 'querystring'
+
+import {
+  authenticateUser,
+  getUserWithBearerToken
+} from '../../../middlewares/authenticateUser'
+import { validateRequest } from '../../../middlewares/validateRequest'
+import OAuth from '../../../models/OAuth'
+import User from '../../../models/User'
+import {
+  expiresIn,
+  generateAccessToken,
+  generateRefreshToken
+} from '../../../utils/config/jwtToken'
+import { ForbiddenError } from '../../../utils/errors/ForbiddenError'
+import { buildQueryURL } from '../utils/buildQueryURL'
+import { isValidRedirectURIValidation } from '../utils/isValidRedirectURIValidation'
+
+const PROVIDER = 'discord'
+const DISCORD_BASE_URL = 'https://discordapp.com/api/v6'
+
+const getUserDiscordData = async (
+  code: string,
+  redirectURI: string
+): Promise<{
+  id: string
+  username: string
+  discriminator: string
+}> => {
+  const { data: dataTokens } = await axios.post<{
+    access_token: string
+    token_type: string
+    expires_in: number
+    refresh_token: string
+    scope: 'identify'
+  }>(
+    `${DISCORD_BASE_URL}/oauth2/token`,
+    querystring.stringify({
+      client_id: process.env.DISCORD_CLIENT_ID,
+      client_secret: process.env.DISCORD_CLIENT_SECRET,
+      grant_type: 'authorization_code',
+      code,
+      redirect_uri: redirectURI,
+      scope: 'identify'
+    }),
+    {
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded'
+      }
+    }
+  )
+  const { data: dataUser } = await axios.get(`${DISCORD_BASE_URL}/users/@me`, {
+    headers: {
+      Authorization: `${dataTokens.token_type} ${dataTokens.access_token}`
+    }
+  })
+  return dataUser
+}
+
+const discordRouter = Router()
+
+discordRouter.get(
+  '/add-strategy',
+  authenticateUser,
+  [
+    query('redirectURI')
+      .notEmpty()
+      .trim()
+      .custom(isValidRedirectURIValidation)
+  ],
+  validateRequest,
+  (req: Request, res: Response) => {
+    if (req.user == null) {
+      throw new ForbiddenError()
+    }
+    const { redirectURI } = req.query as { redirectURI: string }
+    const redirectCallback = `${process.env.API_BASE_URL}/users/oauth2/discord/callback-add-strategy?redirectURI=${redirectURI}`
```

Comment:
I think it's best not to hardcode the whole path
```suggestion
    const redirectCallback = `${req.baseUrl}/callback-add-strategy?redirectURI=${redirectURI}`
```

Reply:
I don't know if `req.baseUrl` is also completing the url with `http://localhost:8080` for example I guess it's only doing the path : `/users/oauth2/discord`, is it right ?

Reply:
Right, good point, the domain would be missing 😃